### ISSUE TITLE
Twisted Web broken on Python 2.7.7.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7461]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)| @oberstet reported|
|-|-|
|Trac ID|trac#7461|
|Type|defect|
|Created|2014-06-02 18:17:27Z|

Twisted Web seems to be broken on Python 2.7.7 (tested on Windows 7, CPython 2.7.7 32 Bit - with Twisted 13.2 and 14.0).

When serving static stuff via `twisted.web.static.File`, the MIME types map `File.contentTypes` ends up having (some) values of type unicode. This break further down the line ("data must not be unicode").

To check, run this on Python 2.7.7:

```
from twisted.internet import reactor
from twisted.web.server import Site
from twisted.web.static import File
from pprint import pprint
 
root = File(".")
pprint(root.contentTypes)
site = Site(root)
reactor.listenTCP(8080, site)
reactor.run()
```

I don't know how these values get unicode, but it doesn't happen on Python 2.6.6.

To work around: https://github.com/crossbario/crossbar/blob/master/crossbar/crossbar/twisted/site.py#L24

Attachments:

* [traceback_7461.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c40/c40a31fffcdb7d5a74ff9b33c333b4a21793c2ad/6f3c4ceaae4cac5f6c813928dd80e031cbafe0f5.txt) (16020 bytes) - added by oberstet on 2014-06-02 18:38:48Z - add traceback

<details><summary>Searchable metadata</summary>

```
trac-id__7461 7461
type__defect defect
reporter__oberstet oberstet
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__web web
keywords__ 
time__1401733047000000 1401733047000000
changetime__1410348955505115 1410348955505115
version__None None
owner__ 
cc__jknight cc__oberstet
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

It would be convenient if you attached the complete traceback and exception that you encounter with this example.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_3"></a>|@jyknight commented|
|-|-|

Almost certainly due to this change:
http://hg.python.org/cpython/rev/18cfc2a42772/
http://bugs.python.org/issue9291

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

anyone interested in reporting this regression upstream?

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_5"></a>|@jyknight commented|
|-|-|

Done, http://bugs.python.org/issue21652

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

This is a regression in Python, not a regression in Twisted.  I don't think it *needs* to be a 14.1.0 blocker.  Not that it wouldn't be nice to have it fixed for 14.1.0.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

Replying to [exarkun](#note_7):
> This is a regression in Python, not a regression in Twisted.  I don't think it *needs* to be a 14.1.0 blocker.  Not that it wouldn't be nice to have it fixed for 14.1.0.

This is just so it doesn't slip under the radar when I release 14.1 (might be useful to mention it as a known issue). I don't really think it's a blocker either (due to it being Python's fault) but it does impact Twisted users.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

http://bugs.python.org/issue21652 is now closed.  It would be good if someone re-tested this issue against the Python 2.7 branch@HEAD.

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_10"></a>|@oberstet commented|
|-|-|

I tested CPython 2.7.8 (32 bit, Windows): the issue is gone.

So the issue is only on 2.7.7 / Windows.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl set status to `closed`|
|-|-|

Since we've got a confirmation that it's fixed in a released Python, I'm going to close this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `reopened`|
|-|-|

The issue is real.  If we're not going to fix it, the proper resolution for the ticket is "wontfix".  "invalid" should be used for issues that don't make sense (because of language or semantic problems) or aren't actually issues (eg, questions about how to use twisted).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

To paraphrase hawkowl, since is fixed in Python 2.7.8 we'll leave Twisted as-is and it will just be broken against 2.7.7 indefinitely.  Users should upgrade to 2.7.8.
